File: tests/wasm/test_alphatrend.js

import test from 'node:test';
import assert from 'node:assert';
import path from 'path';
import { fileURLToPath } from 'url';
import fs from 'fs';
import { parse } from 'csv-parse/sync';

const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);

let wasm;
let testData;

test.before(async () => {

    try {
        const wasmPath = path.join(__dirname, '../../pkg/vector_ta.js');
        const importPath = process.platform === 'win32'
            ? 'file:///' + wasmPath.replace(/\\/g, '/')
            : wasmPath;
        wasm = await import(importPath);
    } catch (error) {
        console.error('Failed to load WASM module:', error);
        throw error;
    }


    const csvPath = path.join(__dirname, '../../src/data/2018-09-01-2024-Bitfinex_Spot-4h.csv');
    try {
        const csvContent = fs.readFileSync(csvPath, 'utf-8');
        const records = parse(csvContent, {
            columns: false,
            skip_empty_lines: true
        });


        testData = {
            open: new Float64Array(records.map(r => parseFloat(r[1]))),
            high: new Float64Array(records.map(r => parseFloat(r[3]))),
            low: new Float64Array(records.map(r => parseFloat(r[4]))),
            close: new Float64Array(records.map(r => parseFloat(r[2]))),
            volume: new Float64Array(records.map(r => parseFloat(r[5])))
        };
    } catch (error) {
        console.error('Failed to load test data:', error);
        throw error;
    }
});

test('AlphaTrend accuracy', async () => {
    const expected = wasm.EXPECTED_OUTPUTS?.alphatrend || {
        defaultParams: { coeff: 1.0, period: 14, noVolume: false },
        k1Last5Values: [
            60243.00,
            60243.00,
            60138.92857143,
            60088.42857143,
            59937.21428571
        ],
        k2Last5Values: [
            60542.42857143,
            60454.14285714,
            60243.00,
            60243.00,
            60138.92857143
        ]
    };

    const result = wasm.alphatrend_js(
        testData.open,
        testData.high,
        testData.low,
        testData.close,
        testData.volume,
        expected.defaultParams.coeff,
        expected.defaultParams.period,
        expected.defaultParams.noVolume
    );


    const len = testData.close.length;
    assert.strictEqual(result.cols, len);
    assert.strictEqual(result.rows, 2);
    assert.strictEqual(result.values.length, len * 2);

    const k1 = result.values.slice(0, len);
    const k2 = result.values.slice(len);


    const k1_last5 = k1.slice(-5);
    for (let i = 0; i < 5; i++) {
        if (!isNaN(k1_last5[i])) {
            const diff = Math.abs(k1_last5[i] - expected.k1Last5Values[i]);
            assert.ok(
                diff < 1e-6,
                `K1 mismatch at index ${i}: got ${k1_last5[i]}, expected ${expected.k1Last5Values[i]}`
            );
        }
    }


    const k2_last5 = k2.slice(-5);
    for (let i = 0; i < 5; i++) {
        if (!isNaN(k2_last5[i])) {
            const diff = Math.abs(k2_last5[i] - expected.k2Last5Values[i]);
            assert.ok(
                diff < 1e-6,
                `K2 mismatch at index ${i}: got ${k2_last5[i]}, expected ${expected.k2Last5Values[i]}`
            );
        }
    }
});

test('AlphaTrend with RSI (no_volume=true)', () => {
    const result = wasm.alphatrend_js(
        testData.open,
        testData.high,
        testData.low,
        testData.close,
        testData.volume,
        1.0,
        14,
        true
    );

    const len = testData.close.length;
    assert.strictEqual(result.cols, len);
    assert.strictEqual(result.rows, 2);
    assert.strictEqual(result.values.length, len * 2);

    const k1 = result.values.slice(0, len);
    const k2 = result.values.slice(len);


    const nonNanK1 = k1.filter(v => !isNaN(v));
    const nonNanK2 = k2.filter(v => !isNaN(v));
    assert.ok(nonNanK1.length > 0, 'K1 should have non-NaN values');
    assert.ok(nonNanK2.length > 0, 'K2 should have non-NaN values');
});

test('AlphaTrend zero period', () => {
    const open = new Float64Array([10.0, 20.0, 30.0]);
    const high = new Float64Array([12.0, 22.0, 32.0]);
    const low = new Float64Array([8.0, 18.0, 28.0]);
    const close = new Float64Array([11.0, 21.0, 31.0]);
    const volume = new Float64Array([100.0, 200.0, 300.0]);

    assert.throws(() => {
        wasm.alphatrend_js(open, high, low, close, volume, 1.0, 0, false);
    }, /Invalid period/);
});

test('AlphaTrend empty input', () => {
    const empty = new Float64Array([]);

    assert.throws(() => {
        wasm.alphatrend_js(empty, empty, empty, empty, empty, 1.0, 14, false);
    }, /empty/i);
});

test('AlphaTrend inconsistent lengths', () => {
    const open = new Float64Array([10.0, 20.0, 30.0]);
    const high = new Float64Array([12.0, 22.0]);
    const low = new Float64Array([8.0, 18.0, 28.0]);
    const close = new Float64Array([11.0, 21.0, 31.0]);
    const volume = new Float64Array([100.0, 200.0, 300.0]);

    assert.throws(() => {
        wasm.alphatrend_js(open, high, low, close, volume, 1.0, 14, false);
    }, /Inconsistent data lengths/);
});

test('AlphaTrend invalid coefficient', () => {

    const data = new Float64Array(Array(20).fill(0).map((_, i) => 10.0 + i * 5));

    assert.throws(() => {
        wasm.alphatrend_js(data, data, data, data, data, -1.0, 14, false);
    }, /Invalid coefficient/);
});

test('AlphaTrend period exceeds length', () => {
    const smallData = new Float64Array([10.0, 20.0, 30.0]);

    assert.throws(() => {
        wasm.alphatrend_js(
            smallData, smallData, smallData, smallData, smallData,
            1.0, 10, false
        );
    }, /Invalid period/);
});

test('AlphaTrend all NaN', () => {
    const nanData = new Float64Array([NaN, NaN, NaN]);

    assert.throws(() => {
        wasm.alphatrend_js(nanData, nanData, nanData, nanData, nanData, 1.0, 14, false);
    }, /All values are NaN/);
});

test('AlphaTrend different parameters', () => {

    const size = 100;
    const open = new Float64Array(size);
    const high = new Float64Array(size);
    const low = new Float64Array(size);
    const close = new Float64Array(size);
    const volume = new Float64Array(size);


    for (let i = 0; i < size; i++) {
        const base = 100 + i * 0.5;
        open[i] = base;
        high[i] = base + 2;
        low[i] = base - 2;
        close[i] = base + Math.sin(i * 0.1);
        volume[i] = 1000 + i * 10;
    }


    const coeffs = [0.5, 1.0, 2.0];
    for (const coeff of coeffs) {
        const result = wasm.alphatrend_js(
            open, high, low, close, volume,
            coeff, 14, false
        );
        assert.strictEqual(result.values.length, size * 2, `Length check for coeff=${coeff}`);
    }


    const periods = [7, 14, 21];
    for (const period of periods) {
        const result = wasm.alphatrend_js(
            open, high, low, close, volume,
            1.0, period, false
        );
        assert.strictEqual(result.values.length, size * 2, `Length check for period=${period}`);
    }


    const resultMFI = wasm.alphatrend_js(
        open, high, low, close, volume,
        1.0, 14, false
    );
    const resultRSI = wasm.alphatrend_js(
        open, high, low, close, volume,
        1.0, 14, true
    );


    const k1_mfi = resultMFI.values.slice(0, size);
    const k1_rsi = resultRSI.values.slice(0, size);

    let foundDifference = false;
    for (let i = 20; i < size; i++) {
        if (!isNaN(k1_mfi[i]) && !isNaN(k1_rsi[i])) {
            if (Math.abs(k1_mfi[i] - k1_rsi[i]) > 1e-10) {
                foundDifference = true;
                break;
            }
        }
    }
    assert.ok(foundDifference, 'MFI and RSI modes should produce different results');
});

test('AlphaTrend memory allocation', () => {
    const size = 1000;
    const ptr = wasm.alphatrend_alloc_flat(size);

    assert.ok(ptr !== 0, 'Should allocate memory successfully');


    wasm.alphatrend_free_flat(ptr, size);
});

test('AlphaTrend warmup period', () => {
    const expected = wasm.EXPECTED_OUTPUTS?.alphatrend || {
        defaultParams: { coeff: 1.0, period: 14, noVolume: false },
        warmupPeriod: 13
    };

    const result = wasm.alphatrend_js(
        testData.open,
        testData.high,
        testData.low,
        testData.close,
        testData.volume,
        expected.defaultParams.coeff,
        expected.defaultParams.period,
        expected.defaultParams.noVolume
    );

    const len = testData.close.length;
    const k1 = result.values.slice(0, len);
    const k2 = result.values.slice(len);
    const warmup = expected.warmupPeriod;


    for (let i = 0; i < warmup && i < k1.length; i++) {
        assert.ok(isNaN(k1[i]), `Expected NaN in K1 warmup at index ${i}`);
        assert.ok(isNaN(k2[i]), `Expected NaN in K2 warmup at index ${i}`);
    }


    if (k1.length > warmup + 10) {
        let foundNonNaN = false;
        for (let i = warmup; i < warmup + 10; i++) {
            if (!isNaN(k1[i])) {
                foundNonNaN = true;
                break;
            }
        }
        assert.ok(foundNonNaN, 'K1 should have real values after warmup');

        foundNonNaN = false;
        for (let i = warmup; i < warmup + 10; i++) {
            if (!isNaN(k2[i])) {
                foundNonNaN = true;
                break;
            }
        }
        assert.ok(foundNonNaN, 'K2 should have real values after warmup');
    }
});

test('AlphaTrend batch processing', () => {

    if (typeof wasm.alphatrend_batch !== 'function') {
        console.log('AlphaTrend batch function not yet exposed in WASM bindings');

        return;
    }


    const size = Math.min(200, testData.close.length);
    const open = new Float64Array(testData.open.slice(0, size));
    const high = new Float64Array(testData.high.slice(0, size));
    const low = new Float64Array(testData.low.slice(0, size));
    const close = new Float64Array(testData.close.slice(0, size));
    const volume = new Float64Array(testData.volume.slice(0, size));




    const singleResult = wasm.alphatrend_batch(
        open, high, low, close, volume,
        1.0, 1.0, 0,
        14, 14, 0,
        false
    );

    assert.ok(singleResult.values, 'Batch result should have values');



    assert.strictEqual(singleResult.rows, 2, 'Should have 2 rows for single combo (K1 and K2)');
    assert.strictEqual(singleResult.cols, size, 'Should have correct column count');


    assert.strictEqual(singleResult.values.length, 2 * size, 'Single combo should have 2 * size values');


    const multiResult = wasm.alphatrend_batch(
        open, high, low, close, volume,
        0.5, 2.0, 0.5,
        7, 21, 7,
        false
    );



    const expectedRows = 24;
    assert.strictEqual(
        multiResult.rows,
        expectedRows,
        `Should have ${expectedRows} rows (12 combos * 2 for K1 and K2)`
    );


    assert.strictEqual(
        multiResult.values.length,
        expectedRows * size,
        'Batch values should have correct total length'
    );
});

test('AlphaTrend NaN distribution', () => {

    const size = Math.min(300, testData.close.length);
    const open = new Float64Array(testData.open.slice(0, size));
    const high = new Float64Array(testData.high.slice(0, size));
    const low = new Float64Array(testData.low.slice(0, size));
    const close = new Float64Array(testData.close.slice(0, size));
    const volume = new Float64Array(testData.volume.slice(0, size));


    const nanIndices = [50, 100, 150, 200, 250].filter(i => i < size);
    for (const idx of nanIndices) {
        close[idx] = NaN;
        high[idx] = NaN;
        low[idx] = NaN;
    }


    const result = wasm.alphatrend_js(
        open, high, low, close, volume,
        1.0, 14, false
    );

    const k1 = result.values.slice(0, size);
    const k2 = result.values.slice(size);

    assert.strictEqual(k1.length, size, 'K1 should have correct length');
    assert.strictEqual(k2.length, size, 'K2 should have correct length');


    for (const idx of nanIndices) {

        assert.ok(
            isNaN(k1[idx]) || isNaN(k2[idx]),
            `Expected NaN propagation at index ${idx}`
        );
    }
});

test('AlphaTrend with real market conditions', () => {

    const size = 50;
    const open = new Float64Array(size);
    const high = new Float64Array(size);
    const low = new Float64Array(size);
    const close = new Float64Array(size);
    const volume = new Float64Array(size);


    for (let i = 0; i < size; i++) {
        const base = 100 + i * 2;
        open[i] = base;
        high[i] = base + 3;
        low[i] = base - 1;
        close[i] = base + 2;
        volume[i] = 1000 + Math.random() * 500;
    }

    const uptrend = wasm.alphatrend_js(
        open, high, low, close, volume,
        1.0, 14, false
    );


    for (let i = 0; i < size; i++) {
        const base = 200 - i * 2;
        open[i] = base;
        high[i] = base + 1;
        low[i] = base - 3;
        close[i] = base - 2;
        volume[i] = 1000 + Math.random() * 500;
    }

    const downtrend = wasm.alphatrend_js(
        open, high, low, close, volume,
        1.0, 14, false
    );


    for (let i = 0; i < size; i++) {
        const base = 150 + Math.sin(i * 0.5) * 5;
        open[i] = base;
        high[i] = base + 2;
        low[i] = base - 2;
        close[i] = base + Math.sin(i * 0.5 + 1) * 1;
        volume[i] = 1000 + Math.random() * 500;
    }

    const sideways = wasm.alphatrend_js(
        open, high, low, close, volume,
        1.0, 14, false
    );


    assert.strictEqual(uptrend.values.length, size * 2);
    assert.strictEqual(downtrend.values.length, size * 2);
    assert.strictEqual(sideways.values.length, size * 2);
});

File: tests/python/test_alphatrend.py
"""
Python binding tests for AlphaTrend indicator.
These tests mirror the Rust unit tests to ensure Python bindings work correctly.
"""
import pytest
import numpy as np
import sys
from pathlib import Path

try:
    import my_project as ta_indicators
except ImportError:
    
    try:
        import my_project as ta_indicators
    except ImportError:
        pytest.skip("Python module not built. Run 'maturin develop --features python' first", allow_module_level=True)

from test_utils import load_test_data, assert_close, EXPECTED_OUTPUTS

class TestAlphaTrend:
    @pytest.fixture(scope='class')
    def test_data(self):
        """Load test data from CSV file"""
        return load_test_data()
    
    def test_alphatrend_accuracy(self, test_data):
        """Test AlphaTrend matches expected values from Rust tests"""
        open_data = test_data['open']
        high = test_data['high']
        low = test_data['low']
        close = test_data['close']
        volume = test_data['volume']
        expected = EXPECTED_OUTPUTS['alphatrend']
        
        
        k1, k2 = ta_indicators.alphatrend(
            open_data,
            high,
            low,
            close,
            volume,
            coeff=expected['default_params']['coeff'],
            period=expected['default_params']['period'],
            no_volume=expected['default_params']['no_volume']
        )
        
        assert len(k1) == len(close)
        assert len(k2) == len(close)
        
        
        
        assert_close(
            k1[-5:],
            expected['k1_last_5_values'],
            rtol=0.0,
            atol=1e-6,
            msg="AlphaTrend K1 last 5 values mismatch"
        )
        
        
        
        assert_close(
            k2[-5:],
            expected['k2_last_5_values'],
            rtol=0.0,
            atol=1e-6,
            msg="AlphaTrend K2 last 5 values mismatch"
        )
    
    def test_alphatrend_with_rsi(self, test_data):
        """Test AlphaTrend with RSI instead of MFI (no_volume=True)"""
        open_data = test_data['open']
        high = test_data['high']
        low = test_data['low']
        close = test_data['close']
        volume = test_data['volume']
        
        
        k1, k2 = ta_indicators.alphatrend(
            open_data,
            high,
            low,
            close,
            volume,
            coeff=1.0,
            period=14,
            no_volume=True
        )
        
        assert len(k1) == len(close)
        assert len(k2) == len(close)
        
        
        non_nan_k1 = k1[~np.isnan(k1)]
        non_nan_k2 = k2[~np.isnan(k2)]
        assert len(non_nan_k1) > 0
        assert len(non_nan_k2) > 0
    
    def test_alphatrend_zero_period(self):
        """Test AlphaTrend fails with zero period"""
        open_data = np.array([10.0, 20.0, 30.0])
        high = np.array([12.0, 22.0, 32.0])
        low = np.array([8.0, 18.0, 28.0])
        close = np.array([11.0, 21.0, 31.0])
        volume = np.array([100.0, 200.0, 300.0])
        
        with pytest.raises(ValueError, match="Invalid period"):
            ta_indicators.alphatrend(open_data, high, low, close, volume, period=0)
    
    def test_alphatrend_empty_input(self):
        """Test AlphaTrend fails with empty input"""
        empty = np.array([])
        
        with pytest.raises(ValueError, match="Input data slice is empty"):
            ta_indicators.alphatrend(empty, empty, empty, empty, empty)
    
    def test_alphatrend_inconsistent_lengths(self):
        """Test AlphaTrend fails with inconsistent array lengths"""
        open_data = np.array([10.0, 20.0, 30.0])
        high = np.array([12.0, 22.0])  
        low = np.array([8.0, 18.0, 28.0])
        close = np.array([11.0, 21.0, 31.0])
        volume = np.array([100.0, 200.0, 300.0])
        
        with pytest.raises(ValueError, match="Inconsistent data lengths"):
            ta_indicators.alphatrend(open_data, high, low, close, volume)
    
    def test_alphatrend_invalid_coeff(self):
        """Test AlphaTrend fails with invalid coefficient"""
        data = np.array([10.0, 20.0, 30.0, 40.0, 50.0] * 5)
        
        with pytest.raises(ValueError, match="Invalid coefficient"):
            ta_indicators.alphatrend(data, data, data, data, data, coeff=-1.0)
    
    def test_alphatrend_period_exceeds_length(self):
        """Test AlphaTrend fails when period exceeds data length"""
        small_data = np.array([10.0, 20.0, 30.0])
        
        with pytest.raises(ValueError, match="Invalid period"):
            ta_indicators.alphatrend(
                small_data, small_data, small_data, small_data, small_data, 
                period=10
            )
    
    def test_alphatrend_all_nan(self):
        """Test AlphaTrend fails with all NaN values"""
        nan_data = np.array([np.nan, np.nan, np.nan])
        
        with pytest.raises(ValueError, match="All values are NaN"):
            ta_indicators.alphatrend(nan_data, nan_data, nan_data, nan_data, nan_data)
    
    def test_alphatrend_stream(self, test_data):
        """Test AlphaTrendStream for streaming calculations"""
        high = test_data['high']
        low = test_data['low']
        close = test_data['close']
        volume = test_data['volume']
        expected = EXPECTED_OUTPUTS['alphatrend']
        
        
        stream = ta_indicators.AlphaTrendStream(
            coeff=expected['default_params']['coeff'],
            period=expected['default_params']['period'],
            no_volume=expected['default_params']['no_volume']
        )
        
        stream_k1 = []
        stream_k2 = []
        
        
        for i in range(min(100, len(close))):  
            result = stream.update(high[i], low[i], close[i], volume[i])
            
            if result is not None:
                stream_k1.append(result[0])
                stream_k2.append(result[1])
            else:
                stream_k1.append(np.nan)
                stream_k2.append(np.nan)
        
        
        
        if all(np.isnan(stream_k1)):
            
            pass
        else:
            
            
            batch_k1, batch_k2 = ta_indicators.alphatrend(
                test_data['open'][:100],
                high[:100],
                low[:100],
                close[:100],
                volume[:100],
                coeff=expected['default_params']['coeff'],
                period=expected['default_params']['period'],
                no_volume=expected['default_params']['no_volume']
            )
            
            
            for i in range(len(stream_k1)):
                if not np.isnan(stream_k1[i]) and not np.isnan(batch_k1[i]):
                    assert_close(stream_k1[i], batch_k1[i], rtol=1e-9,
                                msg=f"K1 streaming mismatch at index {i}")
                if not np.isnan(stream_k2[i]) and not np.isnan(batch_k2[i]):
                    assert_close(stream_k2[i], batch_k2[i], rtol=1e-9,
                                msg=f"K2 streaming mismatch at index {i}")
    
    def test_alphatrend_different_parameters(self):
        """Test AlphaTrend with various parameter combinations"""
        
        size = 100
        np.random.seed(42)
        open_data = np.cumsum(np.random.randn(size)) + 100
        high = open_data + np.abs(np.random.randn(size)) * 2
        low = open_data - np.abs(np.random.randn(size)) * 2
        close = (high + low) / 2 + np.random.randn(size) * 0.5
        volume = np.abs(np.random.randn(size)) * 1000 + 500
        
        
        for coeff in [0.5, 1.0, 2.0]:
            k1, k2 = ta_indicators.alphatrend(
                open_data, high, low, close, volume,
                coeff=coeff, period=14, no_volume=False
            )
            assert len(k1) == size
            assert len(k2) == size
        
        
        for period in [7, 14, 21]:
            k1, k2 = ta_indicators.alphatrend(
                open_data, high, low, close, volume,
                coeff=1.0, period=period, no_volume=False
            )
            assert len(k1) == size
            assert len(k2) == size
        
        
        k1_mfi, k2_mfi = ta_indicators.alphatrend(
            open_data, high, low, close, volume,
            coeff=1.0, period=14, no_volume=False
        )
        k1_rsi, k2_rsi = ta_indicators.alphatrend(
            open_data, high, low, close, volume,
            coeff=1.0, period=14, no_volume=True
        )
        
        
        
        valid_idx = ~(np.isnan(k1_mfi) | np.isnan(k1_rsi))
        if np.any(valid_idx):
            assert not np.allclose(k1_mfi[valid_idx], k1_rsi[valid_idx], rtol=1e-10)
    
    def test_alphatrend_warmup_period(self, test_data):
        """Test AlphaTrend warmup period behavior"""
        close = test_data['close']
        high = test_data['high']
        low = test_data['low']
        volume = test_data['volume']
        expected = EXPECTED_OUTPUTS['alphatrend']
        
        k1, k2 = ta_indicators.alphatrend(
            test_data['open'],
            high,
            low,
            close,
            volume,
            coeff=expected['default_params']['coeff'],
            period=expected['default_params']['period'],
            no_volume=expected['default_params']['no_volume']
        )
        
        
        warmup = expected['warmup_period']
        
        
        assert np.all(np.isnan(k1[:warmup])), f"Expected NaN in K1 warmup period (first {warmup} values)"
        assert np.all(np.isnan(k2[:warmup])), f"Expected NaN in K2 warmup period (first {warmup} values)"
        
        
        if len(k1) > warmup + 10:
            assert not np.all(np.isnan(k1[warmup:warmup+10])), "K1 should have real values after warmup"
            assert not np.all(np.isnan(k2[warmup:warmup+10])), "K2 should have real values after warmup"
    
    @pytest.mark.skip(reason="AlphaTrend batch function not yet implemented in Python bindings")
    def test_alphatrend_batch(self, test_data):
        """Test AlphaTrend batch processing
        
        NOTE: This test is currently skipped because alphatrend_batch is not yet
        implemented in the Python bindings. When implemented, it should follow
        the pattern of other batch functions like alma_batch.
        """
        close = test_data['close'][:200]  
        high = test_data['high'][:200]
        low = test_data['low'][:200]
        open_data = test_data['open'][:200]
        volume = test_data['volume'][:200]
        expected = EXPECTED_OUTPUTS['alphatrend']
        
        
        result = ta_indicators.alphatrend_batch(
            open_data,
            high,
            low,
            close,
            volume,
            coeff_range=(1.0, 1.0, 0.0),
            period_range=(14, 14, 0),
            no_volume=[False]
        )
        
        assert 'k1_values' in result
        assert 'k2_values' in result
        assert 'coeffs' in result
        assert 'periods' in result
        assert 'no_volume_flags' in result
        
        
        assert result['k1_values'].shape[0] == 1
        assert result['k2_values'].shape[0] == 1
        assert result['k1_values'].shape[1] == len(close)
        assert result['k2_values'].shape[1] == len(close)
        
        
        result_multi = ta_indicators.alphatrend_batch(
            open_data,
            high,
            low,
            close,
            volume,
            coeff_range=(0.5, 2.0, 0.5),  
            period_range=(7, 21, 7),       
            no_volume=[False, True]        
        )
        
        
        assert result_multi['k1_values'].shape[0] == 24
        assert result_multi['k2_values'].shape[0] == 24
        
        
        assert len(result_multi['coeffs']) == 24
        assert len(result_multi['periods']) == 24
        assert len(result_multi['no_volume_flags']) == 24
    
    def test_alphatrend_nan_distribution(self, test_data):
        """Test AlphaTrend NaN handling throughout the data"""
        
        close = test_data['close'].copy()
        high = test_data['high'].copy()
        low = test_data['low'].copy()
        open_data = test_data['open'].copy()
        volume = test_data['volume'].copy()
        
        
        nan_indices = [50, 100, 150, 200, 250]
        for idx in nan_indices:
            if idx < len(close):
                close[idx] = np.nan
                high[idx] = np.nan
                low[idx] = np.nan
        
        
        k1, k2 = ta_indicators.alphatrend(
            open_data,
            high,
            low,
            close,
            volume,
            coeff=1.0,
            period=14,
            no_volume=False
        )
        
        assert len(k1) == len(close)
        assert len(k2) == len(close)
        
        
        
        
        for idx in nan_indices:
            if idx < len(k1):
                sticky_ok = False
                if idx > 0:
                    same_k1 = np.isfinite(k1[idx]) and np.isfinite(k1[idx-1]) and np.isclose(k1[idx], k1[idx-1])
                    
                    same_k2 = np.isfinite(k2[idx]) and np.isfinite(k2[idx-1]) and np.isclose(k2[idx], k2[idx-1])
                    sticky_ok = same_k1 or same_k2
                assert (np.isnan(k1[idx]) or np.isnan(k2[idx]) or sticky_ok), \
                    f"Expected NaN propagation or sticky behavior at index {idx}"

if __name__ == "__main__":
    pytest.main([__file__, "-v"])
